toolchain/check/context.h

```diff
@@ -68,37 +69,34 @@ class Context {
   auto AddInst(SemIR::LocIdAndInst loc_id_and_inst) -> SemIR::InstId;
 
   // Convenience for AddInst with typed nodes.
-  template <typename InstT>
-    requires(SemIR::Internal::HasNodeId<InstT>)
-  auto AddInst(decltype(InstT::Kind)::TypedNodeId node_id, InstT inst)
-      -> SemIR::InstId {
-    return AddInst(SemIR::LocIdAndInst(node_id, inst));
+  template <typename InstT, typename LocT>
+  auto AddInst(LocT loc, InstT inst)
+      -> decltype(AddInst(SemIR::LocIdAndInst(loc, inst))) {
+    return AddInst(SemIR::LocIdAndInst(loc, inst));
   }
 
-  // Convenience for AddInst when reusing a location, which any instruction can
-  // do.
+  // Returns a LocIdAndInst for an instruction with an imported location. Checks
+  // that the imported location is compatible with the kind of instruction being
+  // created.
   template <typename InstT>
-  auto AddInstReusingLoc(SemIR::LocId loc_id, InstT inst) -> SemIR::InstId {
-    return AddInst(SemIR::LocIdAndInst::ReusingLoc<InstT>(loc_id, inst));
+    requires SemIR::Internal::HasNodeId<InstT>
+  auto MakeImportedLocAndInst(SemIR::ImportIRInstId imported_loc_id, InstT inst)
+      -> SemIR::LocIdAndInst {
+    if constexpr (!SemIR::Internal::HasUntypedNodeId<InstT>) {
```

Comment:
Out of curiosity, what's an example isntruction that gets an ImportRef and has an untyped node ids? I'm just kind of wondering, because in my head I'm thinking the entities we import would generally have typed node ids, but I expect I'm just forgetting something here.

Reply:
I think `ImportRefUnloaded` and `ImportRefLoaded` are the only ones.

Reply:
Then FWIW, I'll suggest (because it's more explicit, and draws a line in the sand for typed nodes):

```suggestion
    // All other imported instructions should have typed nodes.
    // TODO: Maybe support InstT::Is<AnyImportRef>?
    if constexpr (InstT::Kind != SemIR::ImportRefUnloaded::Kind &&
                  InstT::Kind != SemIR::ImportRefLoaded::Kind) {
```

But understood if you prefer to keep it as-is.

Reply:
I think I'd prefer to keep it as is: I think it's at least slightly happenstance that we have typed nodes for all imported nodes right now. For the most part we're importing declarations where we have a good reason to have typed nodes, but for things like function parameter patterns, I could easily imagine ending up with an instruction with an untyped node, and I think that'd be fine and we shouldn't reject it here.